File: ios-native/Dimo/Domain/ExchangeRates.swift
import Foundation

/// Currency conversion for foreign-currency expense entry. Mirrors the web port
/// in `app/features/currency/rates.ts`. Rates are ECB daily reference rates
/// stored in Convex and refreshed once per day by the `refreshRates` cron —
/// clients read `exchangeRates:latest` and never call Frankfurter themselves.

enum CurrencyMeta {
  struct Info {
    let symbol: String
    let label: String
    let minorUnitDigits: Int
  }

  /// Metadata for every currency a single expense may be entered in.
  static let all: [String: Info] = [
    "INR": Info(symbol: "₹", label: "INR", minorUnitDigits: 2),
    "USD": Info(symbol: "$", label: "USD", minorUnitDigits: 2),
    "EUR": Info(symbol: "€", label: "EUR", minorUnitDigits: 2),
    "GBP": Info(symbol: "£", label: "GBP", minorUnitDigits: 2),
    "JPY": Info(symbol: "¥", label: "JPY", minorUnitDigits: 0),
    "AUD": Info(symbol: "A$", label: "AUD", minorUnitDigits: 2),
    "CAD": Info(symbol: "C$", label: "CAD", minorUnitDigits: 2),
    "HKD": Info(symbol: "HK$", label: "HKD", minorUnitDigits: 2),
    "SGD": Info(symbol: "S$", label: "SGD", minorUnitDigits: 2),
    "CHF": Info(symbol: "CHF", label: "CHF", minorUnitDigits: 2),
    "CNY": Info(symbol: "¥", label: "CNY", minorUnitDigits: 2),
  ]

  /// Ordered list for pickers; default currencies first.
  static let enterable: [String] = [
    "INR", "USD", "EUR", "GBP", "JPY", "AUD", "CAD", "HKD", "SGD", "CHF", "CNY",
  ]

  static func minorUnitDigits(_ code: String) -> Int {
    all[code]?.minorUnitDigits ?? 2
  }

  static func symbol(_ code: String) -> String {
    all[code]?.symbol ?? code
  }

  static func label(_ code: String) -> String {
    all[code]?.label ?? code
  }
}

/// A snapshot of exchange rates for a single day, quoted against `base`.
struct RateTable: Codable, Equatable, Sendable {
  /// ECB rate date, `YYYY-MM-DD`.
  var date: String
  /// Base currency the raw rates were quoted against.
  var base: String
  /// Units of each currency per 1 unit of `base` (the base itself is implicitly 1).
  var rates: [String: Double]
}

enum ExchangeRates {
  private static func factor(_ currency: String) -> Double {
    pow(10.0, Double(CurrencyMeta.minorUnitDigits(currency)))
  }

  /// Major-unit ratio to convert 1 unit of `from` into `to`, or nil if unknown.
  static func rateBetween(_ from: String, _ to: String, _ rates: RateTable?) -> Double? {
    if from == to { return 1 }
    guard let rates else { return nil }
    let unit: (String) -> Double? = { code in code == rates.base ? 1 : rates.rates[code] }
    guard let fromRate = unit(from), let toRate = unit(to), fromRate > 0, toRate > 0 else {
      return nil
    }
    return toRate / fromRate
  }

  /// Convert an integer minor-unit amount between currencies, honoring each
  /// currency's minor-unit exponent. Returns nil when the rate is unavailable.
  static func convertMinor(_ amountMinor: Int, from: String, to: String, rates: RateTable?) -> Int? {
    guard let ratio = rateBetween(from, to, rates) else { return nil }
    let major = (Double(amountMinor) / factor(from)) * ratio
    return Int((major * factor(to)).rounded())
  }

  /// Convert a major-unit amount (what a user types) into minor units.
  static func toMinorUnits(_ amount: Double, _ currency: String) -> Int {
    Int((amount * factor(currency)).rounded())
  }

  /// Convert an integer minor-unit amount back into major units.
  static func toMajorUnits(_ amountMinor: Int, _ currency: String) -> Double {
    Double(amountMinor) / factor(currency)
  }

  /// Canonical recurring fields. New rows always name their denomination.
  static func recurringFields(_ amount: Double, currency: String) -> (amountMinor: Int, currency: String) {
    (max(1, toMinorUnits(amount, currency)), currency)
  }

  /// A recurring bill's amount in major units of `defaultCurrency` using today's
  /// rates. Default-currency bills (or unavailable rates) return the raw amount.
  static func recurringAmountInDefault(
    _ rec: Recurring,
    defaultCurrency: String,
    rates: RateTable?
  ) -> Double {
    guard let currency = rec.currency, currency != defaultCurrency else { return rec.amount }
    let sourceMinor = rec.amountMinor ?? toMinorUnits(rec.amount, currency)
    guard let converted = convertMinor(sourceMinor, from: currency, to: defaultCurrency, rates: rates)
    else { return rec.amount }
    return toMajorUnits(converted, defaultCurrency)
  }
}

/// Local cache for the Convex `exchangeRates:latest` snapshot.
enum RatesService {
  private static let cacheKey = "dimo.exchangeRates"

  static func loadCached() -> RateTable? {
    guard let data = UserDefaults.standard.data(forKey: cacheKey) else { return nil }
    return try? JSONDecoder().decode(RateTable.self, from: data)
  }

  static func store(_ table: RateTable) {
    if let data = try? JSONEncoder().encode(table) {
      UserDefaults.standard.set(data, forKey: cacheKey)
    }
  }
}

File: android-native/app/src/main/java/app/dimo/android/domain/ExchangeRates.kt
package app.dimo.android.domain

import app.dimo.android.data.model.Recurring
import kotlinx.serialization.Serializable
import kotlin.math.max
import kotlin.math.pow
import kotlin.math.roundToLong

/**
 * Port of `ios-native/Dimo/Domain/ExchangeRates.swift`.
 *
 * Rates are ECB daily reference rates stored in Convex and refreshed once per day
 * by the `refreshRates` cron — clients read `exchangeRates:latest` and never call
 * Frankfurter themselves.
 */
object CurrencyMeta {
  data class Info(val symbol: String, val label: String, val minorUnitDigits: Int)

  /** Metadata for every currency a single expense may be entered in. */
  val all: Map<String, Info> = mapOf(
    "INR" to Info("₹", "INR", 2),
    "USD" to Info("$", "USD", 2),
    "EUR" to Info("€", "EUR", 2),
    "GBP" to Info("£", "GBP", 2),
    "JPY" to Info("¥", "JPY", 0),
    "AUD" to Info("A$", "AUD", 2),
    "CAD" to Info("C$", "CAD", 2),
    "HKD" to Info("HK$", "HKD", 2),
    "SGD" to Info("S$", "SGD", 2),
    "CHF" to Info("CHF", "CHF", 2),
    "CNY" to Info("¥", "CNY", 2),
  )

  /** Ordered list for pickers; default currencies first. */
  val enterable: List<String> =
    listOf("INR", "USD", "EUR", "GBP", "JPY", "AUD", "CAD", "HKD", "SGD", "CHF", "CNY")

  fun minorUnitDigits(code: String): Int = all[code]?.minorUnitDigits ?: 2

  fun symbol(code: String): String = all[code]?.symbol ?: code

  fun label(code: String): String = all[code]?.label ?: code
}

/** A snapshot of exchange rates for a single day, quoted against [base]. */
@Serializable
data class RateTable(
  /** ECB rate date, `YYYY-MM-DD`. */
  val date: String,
  /** Base currency the raw rates were quoted against. */
  val base: String,
  /** Units of each currency per 1 unit of [base] (the base itself is implicitly 1). */
  val rates: Map<String, Double>,
)

object ExchangeRates {
  private fun factor(currency: String): Double =
    10.0.pow(CurrencyMeta.minorUnitDigits(currency))

  /** Major-unit ratio to convert 1 unit of [from] into [to], or null if unknown. */
  fun rateBetween(from: String, to: String, rates: RateTable?): Double? {
    if (from == to) return 1.0
    if (rates == null) return null
    fun unit(code: String): Double? = if (code == rates.base) 1.0 else rates.rates[code]
    val fromRate = unit(from) ?: return null
    val toRate = unit(to) ?: return null
    if (fromRate <= 0 || toRate <= 0) return null
    return toRate / fromRate
  }

  /**
   * Convert an integer minor-unit amount between currencies, honoring each
   * currency's minor-unit exponent. Returns null when the rate is unavailable.
   */
  fun convertMinor(amountMinor: Long, from: String, to: String, rates: RateTable?): Long? {
    val ratio = rateBetween(from, to, rates) ?: return null
    val major = (amountMinor.toDouble() / factor(from)) * ratio
    return (major * factor(to)).roundToLong()
  }

  /** Convert a major-unit amount (what a user types) into minor units. */
  fun toMinorUnits(amount: Double, currency: String): Long =
    (amount * factor(currency)).roundToLong()

  /** Convert an integer minor-unit amount back into major units. */
  fun toMajorUnits(amountMinor: Long, currency: String): Double =
    amountMinor.toDouble() / factor(currency)

  /** Canonical recurring fields. New rows always name their denomination. */
  fun recurringFields(amount: Double, currency: String): Pair<Long, String> =
    Pair(max(1L, toMinorUnits(amount, currency)), currency)

  /**
   * A recurring bill's amount in major units of [defaultCurrency] using today's
   * rates. Default-currency bills (or unavailable rates) return the raw amount.
   */
  fun recurringAmountInDefault(
    rec: Recurring,
    defaultCurrency: String,
    rates: RateTable?,
  ): Double {
    val currency = rec.currency
    if (currency == null || currency == defaultCurrency) return rec.amount
    val sourceMinor = rec.amountMinor ?: toMinorUnits(rec.amount, currency)
    val converted = convertMinor(sourceMinor, currency, defaultCurrency, rates) ?: return rec.amount
    return toMajorUnits(converted, defaultCurrency)
  }

  /**
   * A stored transaction's [amountMinor] in major units of [defaultCurrency].
   * Legacy rows without a currency are treated as already in the account default.
   */
  fun transactionAmountInDefault(
    amountMinor: Long,
    currency: String?,
    defaultCurrency: String,
    rates: RateTable?,
  ): Double {
    val code = currency ?: defaultCurrency
    if (code == defaultCurrency) return toMajorUnits(amountMinor, code)
    val converted =
      convertMinor(amountMinor, code, defaultCurrency, rates)
        ?: return toMajorUnits(amountMinor, code)
    return toMajorUnits(converted, defaultCurrency)
  }
}
